internal/operator-controller/authorization/rbac.go

```diff
@@ -0,0 +1,576 @@
+	var (
+		missingRules = map[string][]rbacv1.PolicyRule{}
+		errs         []error
+	)
+	for _, ar := range attributesRecords {
+		allow, err := a.attributesAllowed(ctx, ar)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		if !allow {
+			missingRules[ar.Namespace] = append(missingRules[ar.Namespace], policyRuleFromAttributesRecord(ar))
+		}
+	}
+	return missingRules, errors.Join(errs...)
+}
+
+func (a *rbacPreAuthorizer) attributesAllowed(ctx context.Context, attributesRecord authorizer.AttributesRecord) (bool, error) {
+	decision, reason, err := a.authorizer.Authorize(ctx, attributesRecord)
+	if err != nil {
+		if reason != "" {
+			return false, fmt.Errorf("%s: %w", reason, err)
+		}
+		return false, err
+	}
+	return decision == authorizer.DecisionAllow, nil
+}
+
+func policyRuleFromAttributesRecord(attributesRecord authorizer.AttributesRecord) rbacv1.PolicyRule {
+	pr := rbacv1.PolicyRule{}
+	if attributesRecord.Verb != "" {
+		pr.Verbs = []string{attributesRecord.Verb}
+	}
+	if !attributesRecord.ResourceRequest {
+		pr.NonResourceURLs = []string{attributesRecord.Path}
+		return pr
+	}
+
+	pr.APIGroups = []string{attributesRecord.APIGroup}
+	if attributesRecord.Name != "" {
+		pr.ResourceNames = []string{attributesRecord.Name}
+	}
+
+	r := attributesRecord.Resource
+	if attributesRecord.Subresource != "" {
+		r += "/" + attributesRecord.Subresource
+	}
+	pr.Resources = []string{r}
+
+	return pr
+}
+
+type decodedManifest struct {
+	gvrs                map[schema.GroupVersionResource][]types.NamespacedName
+	clusterRoles        map[client.ObjectKey]rbacv1.ClusterRole
+	roles               map[client.ObjectKey]rbacv1.Role
+	clusterRoleBindings map[client.ObjectKey]rbacv1.ClusterRoleBinding
+	roleBindings        map[client.ObjectKey]rbacv1.RoleBinding
+}
+
+func (dm *decodedManifest) rbacObjects() []client.Object {
+	objects := make([]client.Object, 0, len(dm.clusterRoles)+len(dm.roles)+len(dm.clusterRoleBindings)+len(dm.roleBindings))
+	for obj := range maps.Values(dm.clusterRoles) {
+		objects = append(objects, &obj)
+	}
+	for obj := range maps.Values(dm.roles) {
+		objects = append(objects, &obj)
+	}
+	for obj := range maps.Values(dm.clusterRoleBindings) {
+		objects = append(objects, &obj)
+	}
+	for obj := range maps.Values(dm.roleBindings) {
+		objects = append(objects, &obj)
+	}
+	return objects
+}
+
+func (dm *decodedManifest) asAuthorizationAttributesRecordsForUser(manifestManager user.Info, ext *ocv1.ClusterExtension) []authorizer.AttributesRecord {
+	var attributeRecords []authorizer.AttributesRecord
+
+	// Here we are splitting collection verbs based on required scope
+	// NB: this split is tightly coupled to the requirements of the contentmanager, specifically
+	// its need for cluster-scoped list/watch permissions.
+	// TODO: We are accepting this coupling for now, but plan to decouple
```

Comment:
can we please create a tracking issue for this and link it in this comment?

Reply:
Was done here: https://issues.redhat.com/browse/OPRUN-3823
but I'll make a GH one too, cuz we do that?
https://github.com/operator-framework/operator-controller/issues/1911